File: js/walkway.js
import { LENGTH, createScene } from './environment.js';
import { Agent } from './agent.js';
import { Wall } from './wall.js';
import { updateAgents } from './physics.js'
import * as UTILS from './utils.js'
import * as THREE from 'three';

let agents = [];
let walls = [];
let timestep;

const CONFIG = {
    COUNT: 100,
    RADIUS: 1,
    MAXSPEED: 5,
    MAXFORCE: 20,
    HORIZON: 30,
}

const agentMat = new THREE.MeshLambertMaterial({
    color: 0x00ff00
});

const { renderer, scene, camera } = createScene();
init();

function init() {
    const wallData = [
        [60, 2, true, new THREE.Vector3(0, 1, 15)],
        [60, 2, true, new THREE.Vector3(0, 1, 10)],
        [60, 2, true, new THREE.Vector3(0, 1, -10)],
        [60, 2, true, new THREE.Vector3(0, 1, -15)],
    ];

    wallData.forEach(([width, height, vertical, position]) => {
        const wall = new Wall(width, height, vertical, position);
        scene.add(wall.mesh);
        walls.push(wall);
    });

    const streetMaterial = new THREE.MeshPhongMaterial({ color: 0x222222, side: THREE.DoubleSide });
    const streetGeometry = new THREE.PlaneGeometry(60, 5);

    const street1 = new THREE.Mesh(streetGeometry, streetMaterial);
    street1.castShadow = true;
    street1.receiveShadow = true;
    street1.rotation.set(Math.PI / 2, 0, 0);
    street1.position.set(0, 0.05, 12.5);
    scene.add(street1);

    const street2 = new THREE.Mesh(streetGeometry, streetMaterial);
    street2.castShadow = true;
    street2.receiveShadow = true;
    street2.rotation.set(Math.PI / 2, 0, 0);
    street2.position.set(0, 0.05, -12.5);
    scene.add(street2);

    for (let i = 0; i < CONFIG.COUNT; i++) {
        const v = UTILS.getVelocity(CONFIG.MAXSPEED);
        const pos = UTILS.getPosition(-50, -30, -50, 50);

        const k = 1.5 + Math.random() * 1.5;
        const maxSpeed = Math.random() * (CONFIG.MAXSPEED - 5) + 5;

        agents.push(new Agent(
            i,
            pos[0], 2, pos[1],            
            v[0], 0, v[1],
            0, 0, 0,
            0, 2, pos[1],            
            CONFIG.RADIUS, maxSpeed, CONFIG.MAXFORCE, CONFIG.HORIZON, k
        ));
    }

    agents.forEach(function(member) {
        const agentGeometry = new THREE.CylinderGeometry(member.radius, 1, 4, 16);
        const agentMaterial = new THREE.MeshLambertMaterial({
            color: 0x00ff00
        });
        const agent = new THREE.Mesh(agentGeometry, agentMaterial);
        agent.castShadow = true;
        agent.receiveShadow = true;
        scene.add(agent);
        member.setData("agent", agent);
    });
}

function animate() {
    requestAnimationFrame(animate);

    agents.forEach(function(member) {
        // wrap around
        if (member.position.z > LENGTH / 2 - CONFIG.RADIUS) {
            member.position.z = LENGTH / 2 - CONFIG.RADIUS;
        } else if (member.position.z < -LENGTH / 2 + CONFIG.RADIUS) {
            member.position.z = -LENGTH / 2 + CONFIG.RADIUS;
        }

        // choose left or right walkway
        member.target.z = member.position.z < 0 ? -12.5 : 12.5;
        member.target.x = -32.5;

        // boost speed if in walkway
        if (member.position.x >= -30 && member.position.x <= 30) {
            if (member.position.z < 0 && member.position.z > -15 && member.position.z < -10) {
                member.maxSpeed = CONFIG.MAXSPEED * 1.5;
            } else if (member.position.z > 10 && member.position.z < 15) {
                member.maxSpeed = CONFIG.MAXSPEED * 1.5;
            }
        }

        // funnel agents through walkway
        const nearEntry = Math.abs(member.position.x + 32.5) <= 7.5;
        const left = member.position.z > 10 && member.position.z < 15;
        const right = member.position.z > -15 && member.position.z < -10;
        if ((nearEntry && (left || right)) || (member.position.x >= -30 && (left || right)) ||  member.position.x >= 25) {
            member.target.x = 200;
        }
    });

    timestep = document.getElementById("timestep").value;
    document.getElementById("timestepValue").innerHTML = timestep;
    agents.forEach(function(member) {
        updateAgents(member, agents, timestep);
    });

    agents.forEach(function(agent) {
        walls.forEach(function(wall) {
            wall.collisionResolve(agent);
        });
    });

    agents.forEach(function(member) {
        member.getData("agent").position.copy(member.position);
        member.getData('agent').material = agentMat;
    });

    renderer.render(scene, camera);
};

animate();

File: py/walkway.py
import pygame
from pygame.math import Vector2
from agent import Agent
from wall import Wall
from physics import update_agent
import utils
import random

CONFIG = {
    "COUNT": 75,
    "RADIUS": 3,
    "MAXSPEED": 15,
    "MAXFORCE": 90,
    "HORIZON": 15,
    "K": 2,
    "AVOID": 15,
    "SIDESTEP": 15,
}

SCREEN_WIDTH = 600
SCREEN_HEIGHT = 600
LENGTH = 600

pygame.init()
screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
clock = pygame.time.Clock()

agents = []
walls = []
positions = {}
frame = 0
pause = False
timestep = 0.016

walls_data = [
    (180, True, Vector2(300, 345)),
    (180, True, Vector2(300, 330)),
    (180, True, Vector2(300, 270)),
    (180, True, Vector2(300, 255)),
]

for length, vertical, pos in walls_data:
    walls.append(Wall(length, vertical, pos))

for i in range(CONFIG["COUNT"]):
    pos = utils.get_position(150, 210, 150, 450)
    max_speed = random.uniform(15, CONFIG["MAXSPEED"])
    agents.append(Agent(i, pos, CONFIG["RADIUS"], max_speed, CONFIG["MAXFORCE"],
                        CONFIG["HORIZON"], CONFIG["K"], CONFIG["AVOID"], CONFIG["SIDESTEP"]))

running = True
while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
        elif event.type == pygame.KEYDOWN:
            if event.key == pygame.K_SPACE:
                pause = not pause

    if not pause:
        for agent in agents:
            if agent.position.y > LENGTH - CONFIG["RADIUS"]:
                agent.position.y = LENGTH - CONFIG["RADIUS"]
            elif agent.position.y < CONFIG["RADIUS"]:
                agent.position.y = CONFIG["RADIUS"]

            if agent.position.y > 300:
                agent.target.y = 337.5
            else:
                agent.target.y = 262.5
            agent.target.x = 202.5

            if 210 <= agent.position.x <= 390:
                if 300 < agent.position.y < 345:
                    agent.max_speed = CONFIG["MAXSPEED"] * 1.5
                elif 255 < agent.position.y < 300:
                    agent.max_speed = CONFIG["MAXSPEED"] * 1.5
                else:
                    agent.max_speed = random.uniform(15, CONFIG["MAXSPEED"])

            near_entry = abs(agent.position.x - 202.5) <= 22.5
            left = 330 < agent.position.y < 345
            right = 255 < agent.position.y < 270
            
            if (near_entry and (left or right)) or (agent.position.x >= 210 and (left or right)) or agent.position.x >= 375:
                agent.target.x = 900

        for agent in agents:
            update_agent(agent, agents, timestep)

        for agent in agents:
            for wall in walls:
                wall.collision_resolve(agent)

        frame += 1
        positions[frame] = {agent.id: (agent.position.x, agent.position.y) for agent in agents}

    screen.fill((30, 30, 30))
    for wall in walls:
        pygame.draw.rect(screen, (200, 200, 200), wall.rect)
    for agent in agents:
        pygame.draw.circle(screen, (0, 255, 0), (int(agent.position.x), int(agent.position.y)), int(agent.radius))

    pygame.display.flip()
    clock.tick(60)

pygame.quit()